### ISSUE TITLE
Allowed Clearnet resource switch

### ISSUE THREAD

**Issue by purplesyringa** [HelloZeroNet/ZeroNet#2643]
**Is your feature request related to a problem? Please describe.**

Copied from [this ZeroTalk thread](http://127.0.0.1:43110/Talk.ZeroNetwork.bit/?Topic:1602120041_114yCZGnUqTUcasBbAKUS6xmhQyDPC1ZXt/Iframe+insertion+How+does+TorrentFreak+zite+does+that):

> So... I was surfing around and watching variety of zites at Sites.ZeroNetwork.bit.
>
> Then, I've stumbled upon this zite: http://127.0.0.1:43110/1JNqdTGVATFWRLzzYwVu19CuWYus5VmoUS/
>
> This is TorrentFreak's direct iframe. I mean, this is the fully operational website inside the ZeroNet frame, inserted there by another iframe. In user's data folder it simply looks like this:
>
> ```
> <iframe src="https://torrentfreak.com/" style="position:fixed; top:0px; left:0px; bottom:0px; right:0px; width:100%; height:100%; border:none; margin:0; padding:0; overflow:hidden; z-index:999999;"></iframe>
> ```
>
> Now, if you can allow any website to be rendered from the inside of zeronet iframe, it can possibly lead to some point of anonimity breach, no? I thought, that if you're inside the zeronet iframe, you should not be able to include any other iframes? It's just you can include anything as an iframe in your zite (and hide it), and the end-user won't even know it is there. The hidden iframe can do some malicious things (like monero mining, i.e.), and the page shown could be usual zite to held end-user's attention long enough.
>
> Why iframing external http/https resources is allowed?
>
> -- smashbuckler

**Describe the solution you'd like**
A switch on /Config:

**Allow sites to use Clearnet resources:**
- Yes
- Yes, but warn me when they do
- No

A whitelist should probably be implemented as well.

**Additional context**
This can probably be easily controlled with CSP: the current value works for the 'Yes' option, more limits can be added for 'No', and `Content-Security-Policy-Report-Only` should be used for warnings.

According to some random site on the internet, CSP value limit is around 2048 bytes which should be more than enough for our use cases, including potential URL whitelists.

**Comment by JabbaDesilijicTiure** [comment]
@imachug 
> So... I was surfing around and watching variety of zites at Sites.ZeroNetwork.bit.

You little spy. :rofl: Next time call them sites (the subdomain is **sites**.zeronetwork.but not zites.) better stop using zites.

I just downloaded the site and I think this is absolutely **not okay**! If someone want to have a mirror than create the mirror instead of playing in this network with iframes. Having iframed a clearnet site in this network actually makes the entire network totally useless. The entire point of this network to be decentralised and yet here we are iframing clearnet sites.

In the other hand a clearnet site with `X-Frame-Options: deny` or a `Content-Security-Policy: frame-ancestors 'none'` pretty much end this party... :roll_eyes:

**Comment by JabbaDesilijicTiure** [comment]
@imachug 

I just downloaded the site and I think this is absolutely **not okay**! If someone want to have a mirror than create the mirror instead of playing in this network with iframes. Having iframed a clearnet site in this network actually makes the entire network totally useless. The entire point of this network to be decentralised and yet here we are iframing clearnet sites.

In the other hand a clearnet site with `X-Frame-Options: deny` or a `Content-Security-Policy: frame-ancestors 'none'` pretty much end this party... :roll_eyes: 

This **must be prevented** instead of adding option to /Config..

**Comment by purplesyringa** [comment]
You didn't get the idea -- it's not about iframes in particular, it's about using Clearnet resources like images, scripts, APIs, etc. It looks like I didn't underline that in the issue body though -- sorry for that.

**Comment by JabbaDesilijicTiure** [comment]
No. I'm absolutely against of seeing any site on this network which basically mirroring entire websites. This is far beyond of loading some innocent images. 
<iframe src="https://web.archive.org/web/20180313151740/http://n41050z.web.fc2.com/burakura.html" style="position:fixed; top:0px; left:0px; bottom:0px; right:0px; width:100%; height:100%; border:none; margin:0; padding:0; overflow:hidden; z-index:999999;"></iframe>

**Comment by JabbaDesilijicTiure** [comment]
No. I'm absolutely against of seeing any site on this network which basically mirroring entire websites. This is far beyond of loading some innocent images. 

`<iframe src="https://web.archive.org/web/20180313151740/http://n41050z.web.fc2.com/burakura.html" style="position:fixed; top:0px; left:0px; bottom:0px; right:0px; width:100%; height:100%; border:none; margin:0; padding:0; overflow:hidden; z-index:999999;"></iframe>`

**Don't open in mobile if you create a site with this iframe!**

**Comment by purplesyringa** [comment]
Once again: this issue is not about mirroring entire websites; it's about ZeroNet-to-Clearnet communication.

**Comment by JabbaDesilijicTiure** [comment]
@imachug This is more important: 

Deleting site from data folder results in **hundreds of not found errors** at next start:

```
FileNotFoundError: [Errno 2] No such file or directory: './data/1MaiL5gfBM1cyb4a8e3iiL8L5gXmoAJu27/content.json'
[20:26:36] - Unhandled exception: FileNotFoundError: [Errno 2] No such file or directory: './data/1MaiL5gfBM1cyb4a8e3iiL8L5gXmoAJu27/data/users/content.json' in threadpool.py line 142 > SiteStorage.py line 421 > 290
Traceback (most recent call last):
  File "./lib/python3.8/site-packages/gevent/threadpool.py", line 142, in __run_task
    thread_result.set(func(*args, **kwargs))
  File "./src/Site/SiteStorage.py", line 421, in loadJson
    with self.open(inner_path, "r", encoding="utf8") as file:
  File "./src/Site/SiteStorage.py", line 290, in open
    return open(file_path, mode, **kwargs)
FileNotFoundError: [Errno 2] No such file or directory: './data/1MaiL5gfBM1cyb4a8e3iiL8L5gXmoAJu27/data/users/content.json'
[20:26:36] - Unhandled exception: FileNotFoundError: [Errno 2] No such file or directory: './data/1TaLkFrMwvbNsooF4ioKAY9EuxTBTjipT/content.json' in threadpool.py line 142 > SiteStorage.py line 421 > 290
Traceback (most recent call last):
  File "./lib/python3.8/site-packages/gevent/threadpool.py", line 142, in __run_task
    thread_result.set(func(*args, **kwargs))
  File "./src/Site/SiteStorage.py", line 421, in loadJson
    with self.open(inner_path, "r", encoding="utf8") as file:
  File "./src/Site/SiteStorage.py", line 290, in open
    return open(file_path, mode, **kwargs)
FileNotFoundError: [Errno 2] No such file or directory: './data/1TaLkFrMwvbNsooF4ioKAY9EuxTBTjipT/content.json'
[20:26:36] - Unhandled exception: FileNotFoundError: [Errno 2] No such file or directory: './data/1MaiL5gfBM1cyb4a8e3iiL8L5gXmoAJu27/content.json' in threadpool.py line 142 > SiteStorage.py line 421 > 290
Traceback (most recent call last):
  File "./lib/python3.8/site-packages/gevent/threadpool.py", line 142, in __run_task
    thread_result.set(func(*args, **kwargs))
  File "./src/Site/SiteStorage.py", line 421, in loadJson
    with self.open(inner_path, "r", encoding="utf8") as file:
  File "./src/Site/SiteStorage.py", line 290, in open
    return open(file_path, mode, **kwargs)
```

There should be absolutely no communication with clearnet from this network. Makes no sense to use this software for anything else than for which is designed. Run a server if you want to communicate with clearnet sites...

**Comment by JabbaDesilijicTiure** [comment]
@imachug This is more important: 

Deleting site from data folder results in **hundreds of not found errors** at next start:

```
FileNotFoundError: [Errno 2] No such file or directory: './data/1MaiL5gfBM1cyb4a8e3iiL8L5gXmoAJu27/content.json'
[20:26:36] - Unhandled exception: FileNotFoundError: [Errno 2] No such file or directory: './data/1MaiL5gfBM1cyb4a8e3iiL8L5gXmoAJu27/data/users/content.json' in threadpool.py line 142 > SiteStorage.py line 421 > 290
Traceback (most recent call last):
  File "./lib/python3.8/site-packages/gevent/threadpool.py", line 142, in __run_task
    thread_result.set(func(*args, **kwargs))
  File "./src/Site/SiteStorage.py", line 421, in loadJson
    with self.open(inner_path, "r", encoding="utf8") as file:
  File "./src/Site/SiteStorage.py", line 290, in open
    return open(file_path, mode, **kwargs)
FileNotFoundError: [Errno 2] No such file or directory: './data/1MaiL5gfBM1cyb4a8e3iiL8L5gXmoAJu27/data/users/content.json'
[20:26:36] - Unhandled exception: FileNotFoundError: [Errno 2] No such file or directory: './data/1TaLkFrMwvbNsooF4ioKAY9EuxTBTjipT/content.json' in threadpool.py line 142 > SiteStorage.py line 421 > 290
Traceback (most recent call last):
  File "./lib/python3.8/site-packages/gevent/threadpool.py", line 142, in __run_task
    thread_result.set(func(*args, **kwargs))
  File "./src/Site/SiteStorage.py", line 421, in loadJson
    with self.open(inner_path, "r", encoding="utf8") as file:
  File "./src/Site/SiteStorage.py", line 290, in open
    return open(file_path, mode, **kwargs)
FileNotFoundError: [Errno 2] No such file or directory: './data/1TaLkFrMwvbNsooF4ioKAY9EuxTBTjipT/content.json'
[20:26:36] - Unhandled exception: FileNotFoundError: [Errno 2] No such file or directory: './data/1MaiL5gfBM1cyb4a8e3iiL8L5gXmoAJu27/content.json' in threadpool.py line 142 > SiteStorage.py line 421 > 290
Traceback (most recent call last):
  File "./lib/python3.8/site-packages/gevent/threadpool.py", line 142, in __run_task
    thread_result.set(func(*args, **kwargs))
  File "./src/Site/SiteStorage.py", line 421, in loadJson
    with self.open(inner_path, "r", encoding="utf8") as file:
  File "./src/Site/SiteStorage.py", line 290, in open
    return open(file_path, mode, **kwargs)
```

**There should be absolutely no communication with clearnet from this network**. Makes no sense to use this software for anything else than for which is designed. Run a server if you want to communicate with clearnet sites...

**Comment by JabbaDesilijicTiure** [comment]
Take a look instead to this: https://github.com/HelloZeroNet/ZeroNet/issues/2644#issue-720830107 
This iframe issue must wait for others opinions as well.

**Comment by purplesyringa** [comment]
> Take a look instead to this: [#2644 (comment)](https://github.com/HelloZeroNet/ZeroNet/issues/2644#issue-720830107)

There is no need to fill this issue with useless references to other issues, I'll look at them anyway.

> This iframe issue must wait for others opinions as well.

Oh god, do you get it that it's not about iframes *at all*? It's about scripts, or fonts, or APIs -- but not about iframes in particular.

**Comment by JabbaDesilijicTiure** [comment]
@imachug :rage: 
@HelloZeroNet @shortcutme :point_down: 

Should never load a goddamn thing from clearnet. This is totally against the purposes of the network. 

**Clearnet stuff can disappear or change to bad and than you site will have a meltdown.** 

I'm against anything which is on clearnet. **ZeroNet is a refuge** not a place to invite the **Interpol** and other agencies. 

Lastly, any attempt to make **connections to clearnet should be avoided in order to preserve anonymity**. Even if you use Tor with ZeroNet cloudflare, any American captcha protected service would seriously make your life very difficult. They all blocking Tor. Maintaining anonymity with clearnet requests from ZeroNet for any purpose is irresponsible, dangerous and many people would end up in jail!

**Comment by purplesyringa** [comment]
Wrong. Use case: ZeroNet social network that allows you to add and verify your email/Facebook/etc. This requires a call to an external service.

**Comment by JabbaDesilijicTiure** [comment]
Nobody cares. We have **zeroid.bit** and anyone anytime can create a different service anonymously. We don't need unencrypted email verification originating from ZeroNet nodes and especially we don't need facebook or any other "social network". This can not only result in imprisonment but if it is turns out that someone who runs millchan and other very sensitive ZeroNet sites actually can be even facing serious body harm just by exposing they identity. 

ZeroNet never should allow any call for any external services. If that is going to be implemented ZeroNet than it should be renamed to **Crypt of ZeroNet** (not crypto, burial place). 

If you are on any social network better you watch this: 
**The Social Dilemma** https://www.imdb.com/title/tt11464826/

**Comment by purplesyringa** [comment]
You are using a social network right now - GitHub. I am not advocating for running Facebook on ZeroNet, i.e. a social network with censorship and spying. Linking external accounts, which is my use case, is not mandatory for the main service but is a nice bonus.

**Comment by JabbaDesilijicTiure** [comment]
**Bonus time in jail.** ZeroNet should avoid any external resource to be loaded. If someone playing with iframes in iframes, fine. But to allow loading anything from facebook or any other network including from GitHub is strongly discouraged.

For http://127.0.0.1:43110/1ADQAHsqsie5PBeQhQgjcKmUu3qdPFg6aA a lot of people would end up in jail and this is nothing compared to other sites on ZeroNet. 

Again, no external connection should be allowed ever! In fact iframe in iframe also should be prevented.

**Comment by purplesyringa** [comment]
> For http://127.0.0.1:43110/1ADQAHsqsie5PBeQhQgjcKmUu3qdPFg6aA a lot of people would end up in jail and this is nothing compared to other sites on ZeroNet.

For hosting /tech/? Go on...

> But to allow loading anything from facebook or any other network including from GitHub is strongly discouraged.

Even if that Clearnet site is API that was built for that very zite? Even if the request is `POST /api/verify-email`?.

**Comment by JabbaDesilijicTiure** [comment]
Why you would need to verify email on ZeroNet? You should use https://github.com/lepture/captcha if you want to verify that the user is human and not a bot... Verifying email is too much.

**Comment by purplesyringa** [comment]
Are you willing to admit there are more use cases of email than verifying if a user is a bot? My ideology has always been that bots must have the same rights, if not more powerful, than users, so I'm not going to add captcha anytime soon. I just want to allow my users to attach emails to their accounts, keeping the following invariant: as long as you trust the site owner (me), you can be sure that if you send a email to that address, the right person will receive it.

**Comment by JabbaDesilijicTiure** [comment]
They can attach email without verification, if need to verify that email the user can add a PGP public key (which matching the email in question) to they account and optionally they may sign a message with the PGP Public Key.

**Comment by purplesyringa** [comment]
*You* can do that. And there is nothing wrong in doing that. But there are many people who want to use ZeroNet but have no idea what PGP is. For *them*, email verification is the way.

**Comment by JabbaDesilijicTiure** [comment]
Teach them what is it.. it is your responsibility. I think if someone get that far that running ZeroNet probably learning about PGP is not that hard to do.

**Comment by purplesyringa** [comment]
> I think if someone get that far that running ZeroNet probably learning about PGP is not that hard to do.

You are overestimating humans. Many people come to ZeroNet not because they don't know about alternatives, but because they are too hard to use for them.

**Comment by scsmash3r** [comment]
> @imachug rage
> @HelloZeroNet @shortcutme point_down
> 
> Should never load a goddamn thing from clearnet. This is totally against the purposes of the network.
> 
> **Clearnet stuff can disappear or change to bad and than you site will have a meltdown.**
> 
> I'm against anything which is on clearnet. **ZeroNet is a refuge** not a place to invite the **Interpol** and other agencies.
> 
> Lastly, any attempt to make **connections to clearnet should be avoided in order to preserve anonymity**. Even if you use Tor with ZeroNet cloudflare or any other American captcha protected "service" (which mainly makes profit on selling personal data) would seriously make your life very difficult. They, **all blocking Tor**. Maintaining anonymity with clearnet requests from ZeroNet for any purpose is irresponsible, dangerous and many people would end up in jail!

You downloaded ZeroNet client from clearnet, and you're now posting comments on GitHub, being on clearnet. Yet, you're against resource loading from clearnet. I think it is a good thing to allow to communicate with resources from all over the places on the net (zeronet to clearnet resource request). But it should be a matter of choice for any particular user: it is up to them to allow or to block such requests.

I think, such requests should be blocked by default, and only allowed per zite (if the zite depends on clearnet resource and want to request it) by the client, who is visiting the zite. 

For now, requesting clearnet libraries or chunks of data is a great aid for data to spread, cause inside ZN network itself, in its current state, there often can be 0 peers. With initially blocked access to outer resources, more and more devs will count that moment and will try to migrate their resources fully into ZN network instead just linking them directly (cause it may not work).

**Comment by JabbaDesilijicTiure** [comment]
Saying this @scsmash3r who follows a **sect**, the Venus Project. Your absolutely baseless opinion are not welcome here, you just repeating what @imachug said. It doesn't matter how much you want clearnet switch not going to be implemented in ZeroNet and if Tamas does implement it, I going to release my fork of ZeroNet to the public and encourage people to abandon the use of ZeroNet all together, while I will patch all new PR submitted here. I will not allow ZeroNet to become a shitshow for kids whoi want to lead they trash from the clearnet, actually **neither @imachug nor you want to load your own files from clearnet rather it is about stealing copyrighted material from others and load/use it as your.**
